tools/aie2xclbin/XCLBinGen.cpp

```diff
@@ -0,0 +1,763 @@
+                                          "-Wno-deprecated-declarations"};
+    aieTargetDefines(flags, TK.TargetArch);
+    flags.push_back(incWorkDir);
+    flags.push_back(cdoIncludes);
+    flags.push_back(aietoolsIncludes);
+    SmallString<64> inputFilePath(TK.InstallDir);
+    sys::path::append(inputFilePath, "data", "generated-source", "gen_cdo.cpp");
+    flags.emplace_back(inputFilePath);
+    flags.push_back("-o");
+    flags.emplace_back(genCdoObj);
+    if (runTool(clangppBin, flags, TK.Verbose) != 0) {
+      return moduleOp.emitOpError("failed to compile gen_cdo.o");
+    }
+  }
+
+  // Generate cdo_main.
+  SmallString<64> cdoMainObj(TK.TempDir);
+  sys::path::append(cdoMainObj, "cdo_main.o");
+  {
+    SmallVector<std::string> flags = {"-fPIC", "-c", "-std=c++17"};
+    flags.push_back(incWorkDir);
+    flags.push_back(cdoIncludes);
+    flags.push_back(aietoolsIncludes);
+    SmallString<64> inputFilePath(TK.InstallDir);
+    sys::path::append(inputFilePath, "data", "generated-source",
+                      "cdo_main.cpp");
+    flags.emplace_back(inputFilePath);
+    flags.push_back("-o");
+    flags.emplace_back(cdoMainObj);
+    if (runTool(clangppBin, flags, TK.Verbose) != 0) {
+      return moduleOp.emitOpError("failed to compile gen_cdo.o");
+    }
+  }
+
+  // Generate cdo_main.out
+  SmallString<64> cdoBinary(TK.TempDir);
+  sys::path::append(cdoBinary, "cdo_main");
+  std::string libraryPath = "LD_LIBRARY_PATH=";
+  {
+    SmallVector<std::string> flags;
+
+    std::string cdoLibPathString("-L");
+    SmallString<64> cdoLibPath(TK.InstallDir);
+    sys::path::append(cdoLibPath, "runtime_lib", "x86_64", "xaiengine", "cdo");
+    cdoLibPathString += cdoLibPath;
+    flags.push_back(cdoLibPathString);
+
+    std::string aietoolLibPathString("-L");
+    SmallString<64> aietoolLibPath(TK.AIEToolsDir);
+    sys::path::append(aietoolLibPath, "lib", "lnx64.o");
+    aietoolLibPathString += aietoolLibPath;
+    flags.push_back(aietoolLibPathString);
+
+    flags.push_back("-lxaienginecdo");
+    flags.push_back("-lcdo_driver");
+
+    libraryPath += cdoLibPath;
+    libraryPath += sys::EnvPathSeparator;
+    libraryPath += aietoolLibPath;
+
+    flags.emplace_back(genCdoObj);
+    flags.emplace_back(cdoMainObj);
+    flags.push_back("-o");
+    flags.emplace_back(cdoBinary);
+
+    if (runTool(clangppBin, flags, TK.Verbose) != 0) {
+      return moduleOp.emitOpError("failed to generate cdo_binary");
+    }
+  }
+
+  // Execute the cdo_main binary.
+  {
+    SmallVector<std::string> flags;
+    flags.push_back("--work-dir-path");
+    // CDO tool doesn't join paths correctly :/
+    std::string workDir = TK.TempDir + "/";
+    flags.push_back(workDir);
+    if (runTool(cdoBinary, flags, TK.Verbose,
+                ArrayRef<StringRef>{libraryPath})) {
+      return moduleOp.emitOpError("failed to execute cdo_main binary");
+    }
+  }
+
+  copy->erase();
+  return success();
+}
```

Comment:
so you can pretty safely now replace this with just calls to [AIETranslateToCDODirect](https://github.com/Xilinx/mlir-aie/blob/main/lib/Targets/AIETargetCDODirect.cpp#L723).